[OSDOCS#12598] GA single stack IPv6 for ShiftStack (#85554)
openshift/openshift-docs: modules/installation-configuring-shiftstack-single-ipv6.adoc

// Module included in the following assemblies:
//
// TBD

:_mod-docs-content-type: PROCEDURE
[id="installation-configuring-shiftstack-single-ipv6_{context}"]
= Configuring a cluster with single-stack IPv6 networking

You can create a single-stack IPv6 cluster on {rh-openstack-first}.

.Prerequisites

> * Your {rh-openstack} deployment has an existing IPv6 subnet.

Reviewer: I'd mention here that this ipv6 subnet should be dhcpv6-stateful only. @MaysaMacedo what do you think?

Author: ```suggestion
* Your {rh-openstack} deployment has an existing IPv6 subnet that is DHCPv6-stateful.
```

* DNS is configured for the existing IPv6 subnet.

> * The IPv6 subnet is added to a {rh-openstack} router, and the router is configured to send router advertisements (RAs). 

Reviewer: For clarity, I would include https://github.com/openshift/installer/pull/9323 on this point.

Reviewer: oh this is the same that Maysa commented on https://github.com/openshift/openshift-docs/pull/85554/files#r1901178233. Max, I let you decide where to put it :)

> * The IPv6 subnet is added to a networking service (Neutron) router, and the router is configured to send router advertisements (RAs). 

Reviewer 2: ```suggestion
* The IPv6 subnet is added to a {rh-openstack} router, and the router is configured to send router advertisements (RAs). 
```

> * The {rh-openstack} network accepts a minimum MTU size of 1380 bytes.

Reviewer: @MaysaMacedo We are testing with MTU 1422. Should we put that value here? Where is this 1380 value coming from?

Author: Leaving this. Unresolve if there are changes to make.

Reviewer 2: I think we still needs Ramon input here, if I remember correctly he had to use a bigger MTU

Reviewer: We had to use a bigger MTU in the underlying infrastructure: We needed to increase the MTU in the geneve interfaces while configuring the openstack computes in RHOSO. So I don't think this is really related to this limitation.

IMO, I would put 1442 here as that is the value we have been testing and we will have on the CI.

Author: ```suggestion
* The {rh-openstack} network accepts a minimum MTU size of 1442 bytes.
```

* You created API and ingress virtual IP addresses (VIPs) as ports on the network and included those addresses in the `install-config.yaml` file.

> 

Reviewer: @MaysaMacedo Don't you think we should we mention that a mirror  with ipv6 interface is required to install?

Author: ```suggestion
* You have a mirror registry with an IPv6 interface.
```

.Procedure

> == Creating Network for the Cluster

Reviewer 2: Is this section going to be removed? Just lines 41-43 and the port creation command could stay, in my opinion

> == Creating Network for the Cluster

Author: See updates.

. Create the network and IPv6 subnet:
  [source,bash]
  ----
  $ openstack network create --project <project-name> --share --external --provider-physical-network <physical-network> --provider-network-type flat v6-network
  ----

  [source,bash]
  ----
  $ openstack subnet create --project <project-name> v6-subnet --subnet-range fd2e:6f44:5dd8:c956::/64 --dhcp --dns-nameserver <dns-address> --network v6-network --ip-version 6 --ipv6-ra-mode stateful --ipv6-address-mode stateful
  ----

> . Add the network to the router's external gateway to enable external connectivity and router advertisements:

Reviewer 2: ```suggestion
. As the network used in the example is a provider network, let's add it to the router's external gateway to enable router advertisements:
```

Author: ```suggestion
. If the network is a provider network, as is true in the example, add it to the external gateway of the router to enable external connectivity and router advertisements:
```

Author: @rlobillo As in:
```suggestion
. If the network is a provider network or tenant network, as is true in the example, add it to the external gateway of the router to enable external connectivity and router advertisements:
```
?

  [source,bash]
  ----
  $ openstack router set --external-gateway v6-network <router-id>

>   ----

Author: ```suggestion
  ----
  + 
  You must add any additional IPv6 subnets that are used in the cluster to an {rh-openstack} router to enable router advertisements.
```

>   $ openstack router set --external-gateway v6-network <router-id>

Reviewer 2: can also add the following note:

https://github.com/openshift/installer/commit/5fd62ea137f9782601bac9d551cfbd4aa441cd9f

Author: How 'bout:

> NOTE: You must add any additional IPv6 subnets that are used in the cluster to an {rh-openstack} router to facilitate router advertisements.

Author: https://github.com/openshift/openshift-docs/pull/85554/files#r1936400925

// Module included in the following assemblies:
//
// * installing/installing_openstack/installing-openstack-installer-custom.adoc

:_mod-docs-content-type: PROCEDURE
[id="installation-configuring-shiftstack-single-ipv6_{context}"]
= Configuring a cluster with single-stack IPv6 networking

You can create a single-stack IPv6 cluster on {rh-openstack-first} after you configure your {rh-openstack} deployment.

> IMPORTANT: You cannot convert a dual-stack cluster into a single-stack IPv6 cluster. Additionally, deployments that rely on Fast Datapath are not supported.

Author: At the moment, zero results for "fast data path" in our docs; several exist for "fast datapath". Keeping for now.

.Prerequisites

> * Your {rh-openstack} deployment has an existing IPv6 subnet that is DHCPv6-stateful.

Reviewer 2: ```suggestion
* Your {rh-openstack} deployment has an existing  DHCPv6-stateful IPv6 subnet and a Network that will be used as the Machine Network.
```

Author: ```suggestion
* Your {rh-openstack} deployment has an existing network with a DHCPv6-stateful IPv6 subnet to use as the machine network.
```

* DNS is configured for the existing IPv6 subnet.
* The IPv6 subnet is added to a {rh-openstack} router, and the router is configured to send router advertisements (RAs).
* You added any additional IPv6 subnets that are used in the cluster to an {rh-openstack} router to enable router advertisements.
+

> NOTE: Using an IPv6 SLAAC subnet is not supported due to a known [OpenStack issue](https://bugzilla.redhat.com/show_bug.cgi?id=2304331) that prevents DNS from working.

Reviewer 2: ```suggestion
NOTE: Using an IPv6 SLAAC subnet is not supported as any defined `dns_nameservers` is not enforced by {rh-openstack} Neutron.
```
This is not really an issue, but it's the way it's expected to work according to the last comment on the bz.

Author: ```suggestion
NOTE: Using an IPv6 SLAAC subnet is not supported as any defined `dns_nameservers` servers are not enforced by {rh-openstack} Neutron.
```

* You have a mirror registry with an IPv6 interface.
* The {rh-openstack} network accepts a minimum MTU size of 1380 bytes.

> * You created API and ingress virtual IP addresses (VIPs) as ports on the network and included those addresses in the `install-config.yaml` file.

Reviewer 2: ```suggestion
* You created API and ingress virtual IP addresses (VIPs) as OpenStack Ports on the Machine Network and included those addresses in the `install-config.yaml` file.
```

Author: * You created API and ingress virtual IP addresses (VIPs) as {rh-openstack} ports on the machine network and included those addresses in the `install-config.yaml` file.

Author: ```suggestion
* You created API and ingress virtual IP addresses (VIPs) as {rh-openstack} ports on the machine network and included those addresses in the install-config.yaml file.
```

+
[source,bash]
----
$ openstack port create ingress --network v6-network
----

. After the networking resources are pre-created, deploy a cluster by using an `install-config.yaml` file that reflects your IPv6 network configuration. As an example:
+
[source,yaml]
----
apiVersion: v1
baseDomain: mydomain.test
compute:
- name: worker
  platform:
    openstack:
      type: m1.xlarge
  replicas: 3
controlPlane:
  name: master
  platform:
    openstack:
      type: m1.xlarge
  replicas: 3
metadata:
  name: mycluster
networking:
  machineNetwork:
  - cidr: "fd2e:6f44:5dd8:c956::/64"
  clusterNetwork:
  - cidr: fd01::/48
    hostPrefix: 64
  serviceNetwork:
  - fd02::/112
platform:
  openstack:
    ingressVIPs: ['fd2e:6f44:5dd8:c956::383']
    apiVIPs: ['fd2e:6f44:5dd8:c956::9a']
    controlPlanePort:
      fixedIPs: # <1>
      - subnet:
          name: subnet-v6
      network:
        name: v6-network
imageContentSources: #<2>
- mirrors:
  - <mirror>
  source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
- mirrors:
  - <mirror>
  source: registry.ci.openshift.org/ocp/release
additionalTrustBundle: |
<certificate_of_the_mirror>
----

> <1> Items under the `platform.openstack.controlPlanePort.fixedIPs` and `platform.openstack.controlPlanePort.network` keys can contain either an ID or a name.

Reviewer 2: ```suggestion
<1> Items under the `platform.openstack.controlPlanePort.fixedIPs` and `platform.openstack.controlPlanePort.network` keys can contain an ID and/or a name.
```

Author: ```suggestion
<1> Items under the `platform.openstack.controlPlanePort.fixedIPs` and `platform.openstack.controlPlanePort.network` keys can contain an ID, a name, or both.
```

> <2> The image content sources section contains the mirror details. For more information on configuring a local image registry, see "Creating a mirror registry with mirror registry for Red Hat OpenShift".

Reviewer 2: Suggestion:
can you highlight that `ingressVIPs` `apiVIPs` must contain the addresses of the ports previously generated? I know you already mentioned before, but maybe pointing out here will be more clear.
Also, the `machineNetwork` CIDR should match the CIDR of the subnet that will be used by the machines which is defined under `controlPlanePort`.

Reviewer 2: In the dual-stack docs we have detailed all fields which are relevant for this type of installation, maybe we could do something similar here?  https://docs.openshift.com/container-platform/4.17/installing/installing_openstack/installing-openstack-installer-custom.html#install-osp-deploy-dualstack_installing-openstack-installer-custom

// Module included in the following assemblies:
//
// * installing/installing_openstack/installing-openstack-installer-custom.adoc

:_mod-docs-content-type: PROCEDURE
[id="installation-configuring-shiftstack-single-ipv6_{context}"]
= Configuring a cluster with single-stack IPv6 networking

You can create a single-stack IPv6 cluster on {rh-openstack-first} after you configure your {rh-openstack} deployment.

> IMPORTANT: You cannot convert a dual-stack cluster into a single-stack IPv6 cluster. Additionally, deployments that rely on Fast Datapath are not supported.

Author: ```suggestion
IMPORTANT: You cannot convert a dual-stack cluster into a single-stack IPv6 cluster. Additionally, deployments that rely on fast data path technologies, like OVS-DPDK and SR-IOV, are not supported.
```

> IMPORTANT: You cannot convert a dual-stack cluster into a single-stack IPv6 cluster. Additionally, deployments that rely on Fast Datapath are not supported.

Reviewer 2: I was chatting with @EmilienM and we think it might not be best to mention the no fast data path support as there is some on going work.
@rlobillo WDYT?

Reviewer: Sure, I'm ok with that

> IMPORTANT: You cannot convert a dual-stack cluster into a single-stack IPv6 cluster. Additionally, deployments that rely on fast data path technologies, like OVS-DPDK and SR-IOV, are not supported.

Author: ```suggestion
IMPORTANT: You cannot convert a dual-stack cluster into a single-stack IPv6 cluster.
```

.Prerequisites

* Your {rh-openstack} deployment has an existing network with a DHCPv6-stateful IPv6 subnet to use as the machine network.
* DNS is configured for the existing IPv6 subnet.
* The IPv6 subnet is added to a {rh-openstack} router, and the router is configured to send router advertisements (RAs).
* You added any additional IPv6 subnets that are used in the cluster to an {rh-openstack} router to enable router advertisements.
+

> NOTE: Using an IPv6 SLAAC subnet is not supported because defined `dns_nameservers` servers are not enforced by {rh-openstack} Neutron.

Reviewer 2: ```suggestion
NOTE: Using an IPv6 SLAAC subnet is not supported because any `dns_nameservers` defined are not enforced by {rh-openstack} Neutron.
```

Author: ```suggestion
NOTE: Using an IPv6 SLAAC subnet is not supported because any `dns_nameservers` addresses are not enforced by {rh-openstack} Neutron.
```

@MaysaMacedo Our style guide requires a noun after something like `dns_nameservers`. What do you think?

> NOTE: Using an IPv6 SLAAC subnet is not supported as any defined `dns_nameservers` servers are not enforced by {rh-openstack} Neutron.

Reviewer: SLAAC and DHCPV6-STATELESS

Reviewer 2: @rlobillo have you tried with stateless? I never tried myself, so I can not be certain we would reach the same status here.

Reviewer: @maxwelldb @MaysaMacedo Let me try and will come back to you

Reviewer: The issue happening in slaac seems not to pass on stateless, so I think @maxwelldb you can revert this change back. Sorry for the noise.

I'll keep investigating about dhcpv6-stateless, I still didn't see a successful run on it.

Author: "NOTE: Using an IPv6 SLAAC subnet is not supported because defined `dns_nameservers` servers are not enforced by {rh-openstack} Neutron."

Reviewer: For the record, dhcpv6-stateless installation worked perfectly fine:

```
$ openstack subnet show subnet-ssipv6
+----------------------+----------------------------------------------------------------+
| Field                | Value                                                          |
+----------------------+----------------------------------------------------------------+
| allocation_pools     | fd2e:6f44:5dd8:c956::1-fd2e:6f44:5dd8:c956:ffff:ffff:ffff:ffff |
| cidr                 | fd2e:6f44:5dd8:c956::/64                                       |
| created_at           | 2025-02-06T11:28:49Z                                           |
| description          |                                                                |
| dns_nameservers      | fd2e:6f44:5dd8:c956:f816:3eff:febb:714a                        |
| dns_publish_fixed_ip | None                                                           |
| enable_dhcp          | True                                                           |
| gateway_ip           | fd2e:6f44:5dd8:c956::                                          |
| host_routes          |                                                                |
| id                   | 672a2510-f045-4ad3-860b-01657f18f78d                           |
| ip_version           | 6                                                              |
| ipv6_address_mode    | dhcpv6-stateless                                               |
| ipv6_ra_mode         | dhcpv6-stateless                                               |
| name                 | subnet-ssipv6                                                  |
| network_id           | b28fa993-3801-4ac9-b739-7a05ad4aedbd                           |
| prefix_length        | None                                                           |
| project_id           | 3d1d993d59654056a6294518482511f6                               |
| revision_number      | 0                                                              |
| segment_id           | None                                                           |
| service_types        | None                                                           |
| subnetpool_id        | None                                                           |
| tags                 |                                                                |
| updated_at           | 2025-02-06T12:49:03Z                                           |
+----------------------+----------------------------------------------------------------+
$ oc get clusterversion
NAME      VERSION                              AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.18.0-0.nightly-2025-02-05-033447   True        False         4h52m   Cluster version is 4.18.0-0.nightly-2025-02-05-033447
$ oc get nodes -o wide
NAME                          STATUS   ROLES                  AGE     VERSION   INTERNAL-IP                               EXTERNAL-IP   OS-IMAGE                                                KERNEL-VERSION                 CONTAINER-RUNTIME
ostest-vsrlq-master-0         Ready    control-plane,master   5h31m   v1.31.4   fd2e:6f44:5dd8:c956:f816:3eff:fe80:4ae6   <none>        Red Hat Enterprise Linux CoreOS 418.94.202502041521-0   5.14.0-427.50.1.el9_4.x86_64   cri-o://1.31.5-3.rhaos4.18.git5184b74.el9
ostest-vsrlq-master-1         Ready    control-plane,master   5h33m   v1.31.4   fd2e:6f44:5dd8:c956:f816:3eff:fe8b:fe13   <none>        Red Hat Enterprise Linux CoreOS 418.94.202502041521-0   5.14.0-427.50.1.el9_4.x86_64   cri-o://1.31.5-3.rhaos4.18.git5184b74.el9
ostest-vsrlq-master-2         Ready    control-plane,master   5h33m   v1.31.4   fd2e:6f44:5dd8:c956:f816:3eff:fe1f:9670   <none>        Red Hat Enterprise Linux CoreOS 418.94.202502041521-0   5.14.0-427.50.1.el9_4.x86_64   cri-o://1.31.5-3.rhaos4.18.git5184b74.el9
ostest-vsrlq-worker-0-nxf7z   Ready    worker                 5h4m    v1.31.4   fd2e:6f44:5dd8:c956:f816:3eff:fe15:ca52   <none>        Red Hat Enterprise Linux CoreOS 418.94.202502041521-0   5.14.0-427.50.1.el9_4.x86_64   cri-o://1.31.5-3.rhaos4.18.git5184b74.el9
ostest-vsrlq-worker-0-rmff4   Ready    worker                 5h6m    v1.31.4   fd2e:6f44:5dd8:c956:f816:3eff:fee1:f5c4   <none>        Red Hat Enterprise Linux CoreOS 418.94.202502041521-0   5.14.0-427.50.1.el9_4.x86_64   cri-o://1.31.5-3.rhaos4.18.git5184b74.el9
```

* You have a mirror registry with an IPv6 interface.
* The {rh-openstack} network accepts a minimum MTU size of 1442 bytes.
* You created API and ingress virtual IP addresses (VIPs) as {rh-openstack} ports on the machine network and included those addresses in the `install-config.yaml` file.

.Procedure

. Create the API VIP port on the network by running the following command:
+
[source,bash]
----

> $ openstack port create api --network v6-network

Reviewer 2: ```suggestion
$ openstack port create api --network <v6-machine-network>
```

Author: ```suggestion
$ openstack port create api --network <v6_machine_network>
```

----

. Create the Ingress VIP port on the network by running the following command:
+
[source,bash]
----

> $ openstack port create ingress --network v6-network

Reviewer 2: ```suggestion
$ openstack port create ingress --network <v6-machine-network>
```

----

. After the networking resources are pre-created, deploy a cluster by using an `install-config.yaml` file that reflects your IPv6 network configuration. As an example:
+
[source,yaml]
----
apiVersion: v1
baseDomain: mydomain.test
compute:
- name: worker
  platform:
    openstack:
      type: m1.xlarge
  replicas: 3
controlPlane:
  name: master
  platform:
    openstack:
      type: m1.xlarge
  replicas: 3
metadata:
  name: mycluster
networking:
  machineNetwork:
  - cidr: "fd2e:6f44:5dd8:c956::/64" # <1>
  clusterNetwork:
  - cidr: fd01::/48
    hostPrefix: 64
  serviceNetwork:
  - fd02::/112
platform:
  openstack:
    ingressVIPs: ['fd2e:6f44:5dd8:c956::383'] # <2>
    apiVIPs: ['fd2e:6f44:5dd8:c956::9a'] # <2>
    controlPlanePort:
      fixedIPs: # <3>
      - subnet:
          name: subnet-v6
      network: # <3>
        name: v6-network
imageContentSources: #<4>
- mirrors:
  - <mirror>
  source: quay.io/openshift-release-dev/ocp-v4.0-art-dev
- mirrors:
  - <mirror>
  source: registry.ci.openshift.org/ocp/release
additionalTrustBundle: |
<certificate_of_the_mirror>
----

> <1> The CIDR of the subnet specified in this field must match the CIDR that is specified in the `controlPlanePort` section.

Reviewer 2: ```suggestion
<1> The CIDR of the subnet specified in this field must match the CIDR of the subnet that is specified in the `controlPlanePort` section.
```

> <2> Use the ports you generated in the previous steps as the values for the parameters `platform.openstack.ingressVIPs` and `platform.openstack.apiVIPs`.

Reviewer 2: ```suggestion
<2> Use the address from the ports you generated in the previous steps as the values for the parameters `platform.openstack.ingressVIPs` and `platform.openstack.apiVIPs`.
```

<3> Items under the `platform.openstack.controlPlanePort.fixedIPs` and `platform.openstack.controlPlanePort.network` keys can contain an ID, a name, or both.

> <4> The image content sources section contains the mirror details. For more information on configuring a local image registry, see "Creating a mirror registry with mirror registry for Red Hat OpenShift". 

Reviewer 3: ```suggestion
<4> The `imageContentSources` section contains the mirror details. For more information on configuring a local image registry, see "Creating a mirror registry with mirror registry for Red Hat OpenShift". 
```
